src/book/02-projects/01-nubots/12-mathematics.mdx

```diff
@@ -0,0 +1,158 @@
+---
+section: Projects
+chapter: NUbots
+title: Mathematics
+description: Detailed explanation of the mathematics used in the NUbots codebase.
+slug: /projects/nubots/mathematics
+---
+
+The NUbots codebase uses specific mathematical concepts and notation that is relevant to the field of robotics. This page details these concepts and notations. In the NUbots codebase there are many transforms between spaces. Specific notation is used for these transforms for better readability.
+
+## Bases
+
+A **basis** is a maximal set of linearly independent vectors.
+
+If the vectors in the basis are mutually perpendicular, then they are **orthogonal**.
+
+An **orthonormal basis (ONB)** is a basis which is orthogonal and each vector has length 1.
+
+## Vectors
+
+Any vector $\mathbf{r}$ can be expressed as a linear combination of the elements of a basis:
+
+$$
+\mathbf{r} = r_1^a \mathbf{a_1} + r_2^a \mathbf{a_2} + r_3^a \mathbf{a_3}
+$$
+
+The scalars are the **coordinates** for the vector in the basis $a$.
+
+In the NUbots codebase, vectors are written as `rABb` which denotes a vector from point $B$ to point $A$ in the basis (or **space**) $b$. These are column matrices of the coordinates of the vectors:
+
+$$
+\mathbf{r}^a = \begin{bmatrix}
+r_1^a \\
+r_2^a \\
+r_3^a \end{bmatrix}
+$$
+
+To make computations such as inner product or cross product with coordinates, they must be relative to the same basis.
+
+## Rotations
+
+To describe the motion of the robot, we need a **reference frame** and a **coordinate system**.
+
+The reference frame is the physical rigid body and the coordinate system is a mathematical concept, which is represented as a basis.
+
+**Rotations** are defined by a 3x3 matrix. If we have a vector $\mathbf{r}^a$, that is the vector $r$ in the coordinate system $a$, the rotation
+
+$$
+\begin{bmatrix}
+\cos(\psi) & -\sin(\psi) & 0 \\
+\sin(\psi) & \cos(\psi) & 0 \\
+0 & 0 & 1 \end{bmatrix}
+$$
+
+will rotate $\mathbf{r}^a$ by $\psi$ around the coordinate axis $\mathbf{a_3}$.
+
+The orientation of the basis $b$ relative to $a$ can be described by three consecutive rotations about the main axis. There are 12 different orders to do these rotations in, and each triplet of rotated angles is called a set of **Euler angles**. In robotics, the triplet is **yaw**, **pitch** and **roll** (Z-Y-X).
+
+In the NUbots codebase, we denote a rotation matrix from space $b$ to space $a$ as `Rab`.
+
+There rotation matrices belong to the **special orthogonal group of order 3**. These all have the following properties:
+
+$$
+\mathbf{R}^{-1} = \mathbf{R}^T
+$$
+
+$$
+\text{det}(\mathbf{R}) = 1
+$$
+
+If we have a vector $\mathbf{r}$ (which goes from point C to point D) in a space $a$ and we want to rotate it from space $a$ to space $b$, we write:
+
+$$
+\mathbf{r}^b = \mathbf{R}^b_a \cdot \mathbf{r}^a
```

Comment:
We should add the subscripts to `r` here to tie in the maths notation with the code notation
\mathbf{r}^{b}_{DC} = \mathbf{R}^b_a \cdot \mathbf{r}^{a}_{DC}